volatility3/framework/layers/arm.py

```diff
@@ -0,0 +1,475 @@
+        address."""
+        try:
+            # TODO: Consider reimplementing this, since calls to mapping can call is_valid
+            return all(
+                [
+                    self._context.layers[layer].is_valid(mapped_offset)
+                    for _, _, mapped_offset, _, layer in self.mapping(offset, length)
+                ]
+            )
+        except exceptions.InvalidAddressException:
+            return False
+
+    def is_dirty(self, offset: int) -> bool:
+        """Returns whether the page at offset is marked dirty"""
+        return self._page_is_dirty(self._translate_entry(offset)[2])
+
+    @staticmethod
+    def _page_is_dirty(entry: int) -> bool:
+        """Returns whether a particular page is dirty based on its entry."""
+        # [2], see arch/arm64/include/asm/pgtable-prot.h#L18
+        return bool(entry & (1 << 55))
+
+    @classproperty
+    @functools.lru_cache()
+    def minimum_address(cls) -> int:
+        return 0
+
+    @classproperty
+    @functools.lru_cache()
+    def maximum_address(cls) -> int:
+        return (1 << cls._maxvirtaddr) - 1
+
+    def __canonicalize(self, addr: int) -> int:
+        """Canonicalizes an address by performing an appropiate sign extension on the higher addresses"""
+        if self._bits_per_register <= self._context_maxvirtaddr:
+            return addr & self.address_mask
+        elif addr < (1 << self._context_maxvirtaddr - 1):
+            return addr
+        return self._mask(addr, self._context_maxvirtaddr, 0) + self._canonical_prefix
+
+    def __decanonicalize(self, addr: int) -> int:
+        """Removes canonicalization to ensure an adress fits within the correct range if it has been canonicalized
+
+        This will produce an address outside the range if the canonicalization is incorrect
+        """
+        if addr < (1 << self._context_maxvirtaddr - 1):
+            return addr
+        return addr ^ self._canonical_prefix
+
+    @staticmethod
+    def _mask(
+        value: int, high_bit: int, low_bit: int, shift: Optional[int] = None
+    ) -> int:
+        """Returns the bits of a value between highbit and lowbit inclusive."""
+        if shift is None:
+            shift = low_bit
+        high_mask = (1 << (high_bit + 1)) - 1
+        low_mask = (1 << low_bit) - 1
+        mask = high_mask ^ low_mask
+        return (value & mask) >> shift
+
+    @property
+    def dependencies(self) -> List[str]:
+        """Returns a list of the lower layer names that this layer is dependent
+        upon."""
+        return [self._base_layer]
+
+    @classmethod
+    def get_requirements(cls) -> List[interfaces.configuration.RequirementInterface]:
+        return [
+            requirements.TranslationLayerRequirement(
+                name="memory_layer", optional=False
+            ),
+            requirements.IntRequirement(name="page_map_offset", optional=False),
+            requirements.IntRequirement(name="page_map_offset_kernel", optional=False),
```

Comment:
This... doesn't look like it's used anywhere, and anything related to the kernel shouldn't be a requirement of the layer to operate, it should be optional, because the architecture should be able to map with or without a kernel?

Reply:
`"page_map_offset_kernel"` is instantied once by the LinuxStacker, when creating the kernel layer. This implies that any other caller will try to stack in user space, without passing `"page_map_offset_kernel"` but only `"page_map_offset"` (it is the same naming as Intel, to avoid rewriting everything). Thanks to how the framework is designed, the class will still be able to access `"page_map_offset_kernel"`, which indicates what is the current context (am I being instantiated for user space or for kernel space ?). See : 

```python
self._virtual_addr_space = (
            self._page_map_offset == self.config["page_map_offset_kernel"]
        )
```